src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseGradingService.java

```diff
@@ -333,6 +405,50 @@ private void updateScore(Result result, Set<ProgrammingExerciseTestCase> success
         }
     }
 
+    /**
+     * Calculates the total penalty over all static code analysis issues
+     * @param staticCodeAnalysisFeedback The list of static code analysis feedback
+     * @param programmingExercise The current exercise
+     * @return The sum of all penalties, capped at the maximum allowed penalty
+     */
+    private double calculateStaticCodeAnalysisPenalty(List<Feedback> staticCodeAnalysisFeedback, ProgrammingExercise programmingExercise) {
+
+        double codeAnalysisPenaltyPoints = 0;
+
+        var gradedCategories = staticCodeAnalysisService.findByExerciseId(programmingExercise.getId()).stream()
+                .filter(staticCodeAnalysisCategory -> staticCodeAnalysisCategory.getState().equals(CategoryState.GRADED)).collect(Collectors.toList());
+
+        for (var category : gradedCategories) {
+
+            // get all feedback in this category
+            List<Feedback> categoryFeedback = staticCodeAnalysisFeedback.stream()
+                    .filter(feedback -> feedback.getText().substring(Feedback.STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER.length()).equals(category.getName()))
+                    .collect(Collectors.toList());
```

Comment:
Consider grouping this first (`Collectors.groupingBy`)

```
staticCodeAnalysisFeedback.stream().collect(Collectors.groupingBy(feedback -> feedback.getText().substring(Feedback.STATIC_CODE_ANALYSIS_FEEDBACK_IDENTIFIER.length())))
```

Reply:
Done